### PR TITLE
 Expose Patch operations. Expose missing provider level operations

### PR DESCRIPTION
**Pull request by naveedaz**
This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process. 

### PR information
- [X] The title of the PR is clear and informative.
- [X] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [X] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [X] If applicable, the PR references the bug/issue that it fixes.
- [X] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [X] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [X] My spec meets the review criteria:
  - [X] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [X] The spec follows the guidelines described in the [Swagger checklist](../documentation/swagger-checklist.md) document.
  - [X] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

### CODE REVIEW

**Comment by sergey-shandar** [comment]
@ravbhatnagar the changes contain SDK breaking changes for existing API version. Please, have a look.

**Review by ravbhatnagar — COMMENTED** [review]
@naveedaz - I took a look at this PR. From our offline conversation, the majority of this PR contains adding PATCH support and changing location to be not required. There are some other issues with the models and compliance for which I am not providing feedback since those are existing APIs. But they will need to be fixed at some point. Example include - having top level properties outside of the allowed set instead of nesting them under  "properties". 
A couple of things w.r.t the PATCH change-
1. One problem will happen with the current model of using PATCH. Since you are re-using models between PUT and PATCH, and location is marked as not required, PUT will fail when location is not specified and that is not marked as required.
2. There are new properties that are added for some APIs. This requires an API-version update. Are these getting added to swagger now and were already supported on the service side?
3. Since you are reusing the same model as request body for PUT and PATCH, are all the properties in the request body updatable? If not, I would again recommend to use different model for PATCH.

File: specification/web/resource-manager/Microsoft.Web/2016-03-01/ResourceProvider.json [hunk 1]
```diff
@@ -477,6 +537,233 @@
...
+            "Deleting"
+          ],
+          "type": "string",
+          "readOnly": true,
+          "x-ms-enum": {
+            "name": "HostingEnvironmentStatus",
+            "modelAsString": false
+          }
+        },
+        "vnetName": {
```
[the remark this answers is not in the document]
Comment (naveedaz) [reply specification/web/resource-manager/Microsoft.Web/2016-03-01/ResourceProvider.json]: AppServiceEnvironment is not an ARM Resource. The model 'AppServiceEnvironmentResource' is the ARM Resource.

[the remark this answers is not in the document]
Comment (naveedaz) [reply specification/web/resource-manager/Microsoft.Web/2016-03-01/ResourceProvider.json]: w.r.t PATCH change. 
1. We now have the x-ms-mutability extension on there. Shouldn't that guarantee that location is required for constructor and is read-only otherwise? 
2. Specifically which new properties? We normally change api-version only if there is a breaking change in contract or API behavior. 
3. The patch APIs actually accept the same payloads as PUT since they are internally the same REST API methods. I will have to check if all properties in the payload are patch-able.

**Comment by ravbhatnagar** [comment]
@sergey-shandar - Can you confirm the behavior on #1 above and if SDK team is fine with this?
for #2, I chatted with @naveedaz and the RP behavior is that it does not update properties which are null or not sent on the subsequent PUT. So unintentional update of goal state due to SDK not understanding and ignoring some properties will not happen. 
For #3, yes please check.

**Comment by naveedaz** [comment]
@sergey-shandar, @ravbhatnagar  I generated the SDK and indeed x-ms-mutability does not affect whether property is required. So for now the location will not be required. This is not necessarily breaking but calls from client without location could make it to service, where they will fail with the appropriate message. The validation would move to service instead of client. 
Resource models specifically for patch verb would be a fair bit of work. In the interest of time, I say it is fine to live with the validation moving to service temporarily (customers already specify location when creating so it should not matter much). In a follow up PR I will consolidate common definitions in a single file instead of having it duplicated in all per collection files. It will be much easier to tackle the patching specific resource models then. Discussed this with @ravbhatnagar offline. @sergey-shandar What do you think?

**Comment by ravbhatnagar** [comment]
The proposal by @naveedaz sounds good to me. @sergey-shandar - please confirm fro SDK side.

**Comment by sergey-shandar** [comment]
@shawns1 another PR where CI is not shown. Could you have a look?

File: specification/web/resource-manager/Microsoft.CertificateRegistration/2015-08-01/AppServiceCertificateOrders.json [hunk 2]
```diff
@@ -1341,7 +1465,11 @@
         },
         "location": {
           "description": "Resource Location.",
-          "type": "string"
+          "type": "string",
+          "x-ms-mutability": [
```
Comment (sergey-shandar) [inline specification/web/resource-manager/Microsoft.CertificateRegistration/2015-08-01/AppServiceCertificateOrders.json]: I know, we recommend to use `x-ms-mutability`. AutoRest SDK code generation doesn't use the extension for now but it may use it in the future. It's not an SDK breaking change. In the same time, removing `required` field may change an order of constructor parameters.

File: specification/web/resource-manager/Microsoft.DomainRegistration/2015-04-01/Domains.json [hunk 3]
```diff
@@ -1130,9 +1191,6 @@
     },
     "Resource": {
       "description": "Azure resource. This resource is tracked in Azure Resource Manager",
-      "required": [
-        "location"
```
Comment (sergey-shandar) [inline specification/web/resource-manager/Microsoft.DomainRegistration/2015-04-01/Domains.json]: Same here, it's a potential SDK breaking change.

File: specification/web/resource-manager/Microsoft.Web/2016-03-01/Certificates.json [hunk 4]
```diff
@@ -451,9 +452,6 @@
     },
     "Resource": {
       "description": "Azure resource. This resource is tracked in Azure Resource Manager",
-      "required": [
-        "location"
```
Comment (sergey-shandar) [inline specification/web/resource-manager/Microsoft.Web/2016-03-01/Certificates.json]: A potential SDK breaking change.

File: specification/web/resource-manager/Microsoft.Web/2016-03-01/Provider.json [hunk 5]
```diff
@@ -32,7 +32,7 @@
           "200": {
             "description": "OK",
             "schema": {
-              "$ref": "#/definitions/Object"
+              "type": "object"
```
Comment (sergey-shandar) [inline specification/web/resource-manager/Microsoft.Web/2016-03-01/Provider.json]: an SDK breaking change.

File: specification/web/resource-manager/Microsoft.Web/2016-03-01/Provider.json [hunk 6]
```diff
@@ -92,7 +92,7 @@
...
```
Comment (sergey-shandar) [inline specification/web/resource-manager/Microsoft.Web/2016-03-01/Provider.json]: an SDK breaking change.

File: specification/web/resource-manager/Microsoft.Web/2016-03-01/Recommendations.json [hunk 7]
```diff
@@ -295,6 +295,12 @@
             "type": "boolean"
           },
           {
+            "name": "recommendationId",
+            "in": "query",
```
Comment (sergey-shandar) [inline specification/web/resource-manager/Microsoft.Web/2016-03-01/Recommendations.json]: A new parameter in the middle of parameter list, which is an SDK breaking change and probably Rest API breaking change. @ravbhatnagar. To reduce impact on SDK code generation, the new parameter should be the last one so function signature will not have a breaking change.

File: specification/web/resource-manager/Microsoft.Web/2016-03-01/Recommendations.json [hunk 8]
```diff
@@ -323,8 +329,10 @@
           "type": "string"
         },
         "recommendationId": {
+          "format": "uuid",
           "description": "A GUID value that each recommendation object is associated with.",
-          "type": "string"
+          "type": "string",
```
Comment (sergey-shandar) [inline specification/web/resource-manager/Microsoft.Web/2016-03-01/Recommendations.json]: SDK breaking change.

File: specification/web/resource-manager/Microsoft.Web/2016-08-01/WebApps.json [hunk 9]
```diff
@@ -16706,16 +17010,12 @@
...
-          "description": "ARM resource ID of the Traffic Manager profile to use, if it exists. Traffic Manager resource ID is of the form \n /subscriptions/{subId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/trafficManagerProfiles/{profileName}.",
+          "description": "ARM resource ID of the Traffic Manager profile to use, if it exists. Traffic Manager resource ID is of the form \n/subscriptions/{subId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/trafficManagerProfiles/{profileName}.",
           "type": "string"
         },
         "trafficManagerProfileName": {
           "description": "Name of Traffic Manager profile to create. This is only needed if Traffic Manager profile does not already exist.",
           "type": "string"
-        },
-        "ignoreQuotas": {
-          "description": "<code>true</code> if quotas should be ignored; otherwise, <code>false</code>.",
```
Comment (sergey-shandar) [inline specification/web/resource-manager/Microsoft.Web/2016-08-01/WebApps.json]: The property is removed. SDK and API breaking change.

File: specification/web/resource-manager/Microsoft.Web/2016-08-01/WebApps.json [hunk 10]
```diff
@@ -17264,6 +17556,14 @@
           "items": {
             "$ref": "#/definitions/ErrorEntity"
           }
+        },
+        "code": {
```
Comment (sergey-shandar) [inline specification/web/resource-manager/Microsoft.Web/2016-08-01/WebApps.json]: Do we have a reason to shuffle the properties order?

File: specification/web/resource-manager/Microsoft.Web/2016-08-01/WebApps.json [hunk 11]
```diff
@@ -18256,11 +18558,6 @@
           "format": "double",
           "description": "Value of counter at a certain time.",
           "type": "number"
-        },
-        "coreCount": {
-          "format": "int32",
```
Comment (sergey-shandar) [inline specification/web/resource-manager/Microsoft.Web/2016-08-01/WebApps.json]: The property was removed.

File: specification/web/resource-manager/Microsoft.Web/2016-08-01/WebApps.json [hunk 12]
```diff
@@ -18958,9 +19256,6 @@
     },
     "Resource": {
       "description": "Azure resource. This resource is tracked in Azure Resource Manager",
-      "required": [
-        "location"
```
Comment (sergey-shandar) [inline specification/web/resource-manager/Microsoft.Web/2016-08-01/WebApps.json]: potential SDK breaking change.

File: specification/web/resource-manager/Microsoft.Web/2016-09-01/AppServiceEnvironments.json [hunk 13]
```diff
@@ -1,3006 +1,3222 @@
...
-            "description": "OK",
-            "schema": {
-              "$ref": "#/definitions/AppServiceEnvironmentCollection"
-            }
-          }
-        },
-        "x-ms-pageable": {
-          "nextLinkName": "nextLink"
-        }
+ "swagger": "2.0",
```
Comment (sergey-shandar) [inline specification/web/resource-manager/Microsoft.Web/2016-09-01/AppServiceEnvironments.json]: Changing formatting of the document is unnecessary changes and it's really hard to review. Could you format the document as it was before?

File: specification/web/resource-manager/Microsoft.Web/2016-09-01/AppServicePlans.json [hunk 14]
```diff
@@ -3214,7 +3433,8 @@
               "readOnly": true
             },
             "certBlob": {
-              "description": "A certificate file (.cer) blob containing the public key of the private key used to authenticate a \n Point-To-Site VPN connection.",
+              "format": "byte",
+              "description": "A certificate file (.cer) blob containing the public key of the private key used to authenticate a \nPoint-To-Site VPN connection.",
```
Comment (sergey-shandar) [inline specification/web/resource-manager/Microsoft.Web/2016-09-01/AppServicePlans.json]: A potential SDK breaking changes.

File: specification/web/resource-manager/Microsoft.Web/2016-09-01/AppServicePlans.json [hunk 15]
```diff
@@ -2074,9 +2288,6 @@
     },
     "Resource": {
       "description": "Azure resource. This resource is tracked in Azure Resource Manager",
-      "required": [
-        "location"
-      ],
```
Comment (sergey-shandar) [inline specification/web/resource-manager/Microsoft.Web/2016-09-01/AppServicePlans.json]: A potential SDK breaking changes.

**Review by sergey-shandar — CHANGES_REQUESTED** [review]

File: specification/web/resource-manager/Microsoft.Web/2016-03-01/Recommendations.json (hunk 8 above)
Comment (naveedaz) [reply specification/web/resource-manager/Microsoft.Web/2016-03-01/Recommendations.json]: Client validation for (uuid, byte ) will cause SDK breaks, but is acceptable and actually a good change.

File: specification/web/resource-manager/Microsoft.Web/2016-08-01/WebApps.json (hunk 10 above)
Comment (naveedaz) [reply specification/web/resource-manager/Microsoft.Web/2016-08-01/WebApps.json]: ErrorEntity was incorrectly under properties property. ARM expects it in the envelope.

File: specification/web/resource-manager/Microsoft.Web/2016-03-01/Provider.json (hunk 5 above)
Comment (naveedaz) [reply specification/web/resource-manager/Microsoft.Web/2016-03-01/Provider.json]: Change from #/definitions/Object is SDK but recommended by Azure SDK/ARM

File: specification/web/resource-manager/Microsoft.Web/2016-03-01/Provider.json (hunk 6 above)
Comment (naveedaz) [reply specification/web/resource-manager/Microsoft.Web/2016-03-01/Provider.json]: Change from #/definitions/Object is SDK but recommended by Azure SDK/ARM

File: specification/web/resource-manager/Microsoft.Web/2016-09-01/AppServicePlans.json (hunk 14 above)
Comment (naveedaz) [reply specification/web/resource-manager/Microsoft.Web/2016-09-01/AppServicePlans.json]: Client validation for (uuid, byte ) will cause SDK breaks, but is acceptable and actually a good change.

**Comment by naveedaz** [comment]
1. consumes/produces is now declared only at the collection level unless it is different for a specific operation. 
2. WebJobId was a false query parameter, that should not have existed. The service does not use it. 
3. Client validation for (uuid, byte ) will cause SDK breaks, but is acceptable and actually a good change. 
4. ErrorEntity was incorrectly under properties property. ARM expects it in the envelope. 
5. Change from #/definitions/Object is SDK but recommended by Azure SDK/ARM
6. CommonDefinitions.json is now added. The generator tool just spits the common models in this file. References to this file are not yet used, since it was complicating the PR too much. I will add that as a separate PR.

**Review by sergey-shandar — APPROVED** [review]
There a lot of SDK breaking changes but otherwise the changes are really good.

**Comment by lmazuel** [comment]
@sergey-shandar @naveedaz This PR removed MSI from a loooot of models. Please check asap if it's on purpose.

**Comment by naveedaz** [comment]
MSI is supported only for the site resource. It was intentional to remove it from all tracked resources except for Site. 

Comments from above
7. Resource model does not have identity anymore. Instead it is on the only tracked resource 'Site' that actually uses it. SDK wise the optional property will be removed for all the tracked resources that do not use a managed service identity. (AppServiceCertificateResource, AppServiceCertificateOrder, Certificate
Domai, AppServiceEnvironmentResource, AppServicePlan).

**Comment by lmazuel** [comment]
@naveedaz Yes, I saw afterward and removed my comment (not fast enough :)). Thanks for the confirmation :)

I see ContinuousWebJob.settings has changed type, I guess it's on purpose as well.
